pyquil/noise.py

```diff
@@ -513,3 +513,141 @@ def add_decoherence_noise(prog, T1=30e-6, T2=30e-6, gate_time_1q=50e-9, gate_tim
         ro_fidelity=ro_fidelity
     )
     return _apply_noise_model(prog, noise_model)
+
+
+def _bitstring_probs_by_qubit(p):
+    """
+    Ensure that an array ``p`` with bitstring probabilities has a separate axis for each qubit such
+    that ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+
+    This should not allocate much memory if ``p`` is already in ``C``-contiguous order (row-major).
+
+    :param np.array p: An array that enumerates bitstring probabilities. When flattened out
+        ``p = [p_00...0, p_00...1, ...,p_11...1]``. The total number of elements must therefore be a
+        power of 2.
+    :return: A reshaped view of ``p`` with a separate length-2 axis for each bit.
+    """
+    p = np.asarray(p, order="C")
+    num_qubits = int(round(np.log2(p.size)))
+    return p.reshape((2,) * num_qubits)
+
+
+def estimate_bitstring_probs(results):
+    """
+    Given an array of single shot results estimate the probability distribution over all bitstrings.
+
+    :param np.array results: A 2d array where the outer axis iterates over shots
+        and the inner axis over bits.
+    :return: An array with as many axes as there are qubit and normalized such that it sums to one.
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :rtype: np.array
+    """
+    nshots, nq = np.shape(results)
+    outcomes = np.array([int("".join(map(str, r)), 2) for r in results])
+    probs = np.histogram(outcomes, bins=np.arange(-.5, 2 ** nq, 1))[0] / float(nshots)
+    return _bitstring_probs_by_qubit(probs)
+
+
+_CHARS = 'klmnopqrstuvwxyzabcdefgh'
+
+
+def _apply_local_trafos(p, trafos):
```

Comment:
`trafos` is an unclear variable name.  I guess it means transform.  maybe something like `inv_assignment_errors` or some less verbose version of that.

Reply:
On second thought...maybe `transform` is okay since you use the hadamard -> Z-moment trick later on.

Reply:
Cool, will rename to `transforms`